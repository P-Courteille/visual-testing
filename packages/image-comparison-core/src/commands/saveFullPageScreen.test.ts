import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest'
import saveFullPageScreen from './saveFullPageScreen.js'
import type { InternalSaveFullPageMethodOptions } from './save.interfaces.js'
import {
    BASE_CHECK_OPTIONS,
    createMethodOptions,
    createBeforeScreenshotMock
} from '../mocks/mocks.js'

vi.mock('../helpers/beforeScreenshot.js', () => ({
    default: vi.fn().mockResolvedValue({
        browserName: 'chrome',
        browserVersion: '120.0.0',
        deviceName: 'desktop',
        dimensions: {
            window: {
                devicePixelRatio: 2,
                innerHeight: 900,
                isEmulated: false,
                isLandscape: false,
                outerHeight: 1000,
                outerWidth: 1200,
                screenHeight: 1080,
                screenWidth: 1920,
            },
        },
        isAndroid: false,
        isAndroidChromeDriverScreenshot: false,
        isAndroidNativeWebScreenshot: false,
        isIOS: false,
        isMobile: false,
        isTestInBrowser: true,
        logName: 'chrome',
        name: 'chrome',
        platformName: 'desktop',
        platformVersion: '120.0.0',
    })
}))
vi.mock('../methods/screenshots.js', () => ({
    takeBase64BiDiScreenshot: vi.fn().mockResolvedValue('bidi-screenshot-data'),
    getBase64FullPageScreenshotsData: vi.fn().mockResolvedValue({
        fullPageHeight: 2000,
        fullPageWidth: 1200,
        data: ['screenshot-1', 'screenshot-2']
    })
}))
vi.mock('../methods/images.js', () => ({
    makeFullPageBase64Image: vi.fn().mockResolvedValue('fullpage-screenshot-data')
}))
vi.mock('../helpers/afterScreenshot.js', () => ({
    default: vi.fn().mockResolvedValue({
        devicePixelRatio: 2,
        fileName: 'test-fullpage.png'
    })
}))
vi.mock('../helpers/utils.js', () => ({
    canUseBidiScreenshot: vi.fn().mockReturnValue(false),
    getMethodOrWicOption: vi.fn().mockImplementation((method, wic, option) => {
        if (option === 'userBasedFullPageScreenshot') {return method[option] ?? wic[option]}
        if (option === 'enableLegacyScreenshotMethod') {return method[option] ?? wic[option]}
        return method[option] ?? wic[option]
    })
}))

describe('saveFullPageScreen', () => {
    let takeBase64BiDiScreenshotSpy: ReturnType<typeof vi.fn>
    let getBase64FullPageScreenshotsDataSpy: ReturnType<typeof vi.fn>
    let makeFullPageBase64ImageSpy: ReturnType<typeof vi.fn>
    let afterScreenshotSpy: ReturnType<typeof vi.fn>
    let canUseBidiScreenshotSpy: ReturnType<typeof vi.fn>

    const baseOptions = {
        browserInstance: { isAndroid: false, isMobile: false } as any,
        folders: BASE_CHECK_OPTIONS.folders,
        instanceData: BASE_CHECK_OPTIONS.instanceData,
        isNativeContext: false,
        saveFullPageOptions: {
            wic: BASE_CHECK_OPTIONS.wic,
            method: createMethodOptions({
                disableBlinkingCursor: false,
                disableCSSAnimation: false,
                enableLayoutTesting: false,
                enableLegacyScreenshotMethod: false,
                hideScrollBars: true,
                hideElements: [],
                removeElements: [],
                waitForFontsLoaded: true,
            })
        },
        tag: 'test-fullpage'
    } as InternalSaveFullPageMethodOptions

    const createTestOptions = (methodOptions = {}) => ({
        ...baseOptions,
        saveFullPageOptions: {
            ...baseOptions.saveFullPageOptions,
            method: createMethodOptions({
                ...baseOptions.saveFullPageOptions.method,
                ...methodOptions
            })
        }
    })

    beforeEach(async () => {
        const { takeBase64BiDiScreenshot, getBase64FullPageScreenshotsData } = await import('../methods/screenshots.js')
        const { makeFullPageBase64Image } = await import('../methods/images.js')
        const afterScreenshot = (await import('../helpers/afterScreenshot.js')).default
        const { canUseBidiScreenshot } = await import('../helpers/utils.js')

        takeBase64BiDiScreenshotSpy = vi.mocked(takeBase64BiDiScreenshot)
        getBase64FullPageScreenshotsDataSpy = vi.mocked(getBase64FullPageScreenshotsData)
        makeFullPageBase64ImageSpy = vi.mocked(makeFullPageBase64Image)
        afterScreenshotSpy = vi.mocked(afterScreenshot)
        canUseBidiScreenshotSpy = vi.mocked(canUseBidiScreenshot)
    })

    afterEach(() => {
        vi.clearAllMocks()
    })

    it('should throw an error when in native context', async () => {
        const options = {
            ...baseOptions,
            isNativeContext: true
        }

        await expect(saveFullPageScreen(options)).rejects.toThrow(
            'The method saveFullPageScreen is not supported in native context for native mobile apps!'
        )
    })

    it('should use Bidi screenshot when available and userBasedFullPageScreenshot is false', async () => {
        canUseBidiScreenshotSpy.mockReturnValueOnce(true)
        const options = createTestOptions({
            userBasedFullPageScreenshot: false
        })
        const result = await saveFullPageScreen(options)

        expect(result).toMatchSnapshot()
        expect(takeBase64BiDiScreenshotSpy.mock.calls[0]).toMatchSnapshot()
        expect(getBase64FullPageScreenshotsDataSpy).not.toHaveBeenCalled()
        expect(makeFullPageBase64ImageSpy).not.toHaveBeenCalled()
        expect(afterScreenshotSpy.mock.calls[0]).toMatchSnapshot()
    })

    it('should use Bidi screenshot when available and enableLegacyScreenshotMethod is false', async () => {
        canUseBidiScreenshotSpy.mockReturnValueOnce(true)
        const options = createTestOptions({
            enableLegacyScreenshotMethod: false
        })
        const result = await saveFullPageScreen(options)

        expect(result).toMatchSnapshot()
        expect(takeBase64BiDiScreenshotSpy.mock.calls[0]).toMatchSnapshot()
        expect(getBase64FullPageScreenshotsDataSpy).not.toHaveBeenCalled()
        expect(makeFullPageBase64ImageSpy).not.toHaveBeenCalled()
        expect(afterScreenshotSpy.mock.calls[0]).toMatchSnapshot()
    })

    it('should use legacy method when Bidi is available but both userBasedFullPageScreenshot and enableLegacyScreenshotMethod are true', async () => {
        canUseBidiScreenshotSpy.mockReturnValueOnce(true)
        const options = createTestOptions({
            userBasedFullPageScreenshot: true,
            enableLegacyScreenshotMethod: true
        })
        const result = await saveFullPageScreen(options)

        expect(result).toMatchSnapshot()
        expect(takeBase64BiDiScreenshotSpy).not.toHaveBeenCalled()
        expect(getBase64FullPageScreenshotsDataSpy.mock.calls[0]).toMatchSnapshot()
        expect(makeFullPageBase64ImageSpy.mock.calls[0]).toMatchSnapshot()
        expect(afterScreenshotSpy.mock.calls[0]).toMatchSnapshot()
    })

    it('should handle undefined hideAfterFirstScroll, hideElements, and removeElements', async () => {
        const options = createTestOptions({
            hideAfterFirstScroll: undefined,
            hideElements: undefined,
            removeElements: undefined
        })

        await saveFullPageScreen(options)

        expect(getBase64FullPageScreenshotsDataSpy.mock.calls[0]).toMatchSnapshot()
    })

    it('should handle NaN dimension values', async () => {
        const nanDimensions = createBeforeScreenshotMock({
            dimensions: {
                body: {
                    scrollHeight: NaN,
                    offsetHeight: NaN
                },
                html: {
                    clientWidth: NaN,
                    scrollWidth: NaN,
                    clientHeight: NaN,
                    scrollHeight: NaN,
                    offsetHeight: NaN
                },
                window: {
                    devicePixelRatio: NaN,
                    innerHeight: NaN,
                    isEmulated: false,
                    isLandscape: false,
                    outerHeight: NaN,
                    outerWidth: NaN,
                    screenHeight: NaN,
                    screenWidth: NaN,
                },
            },
            devicePixelRatio: NaN,
            initialDevicePixelRatio: NaN
        })
        vi.mocked((await import('../helpers/beforeScreenshot.js')).default).mockResolvedValueOnce(nanDimensions)
        const result = await saveFullPageScreen(baseOptions)

        expect(result).toMatchSnapshot()
        expect(getBase64FullPageScreenshotsDataSpy.mock.calls[0]).toMatchSnapshot()
        expect(makeFullPageBase64ImageSpy.mock.calls[0]).toMatchSnapshot()
        expect(afterScreenshotSpy.mock.calls[0]).toMatchSnapshot()
    })
})
