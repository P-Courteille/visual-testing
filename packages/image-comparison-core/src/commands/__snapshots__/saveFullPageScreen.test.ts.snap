// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`saveFullPageScreen > should handle NaN dimension values 1`] = `
{
  "devicePixelRatio": 2,
  "fileName": "test-fullpage.png",
}
`;

exports[`saveFullPageScreen > should handle NaN dimension values 2`] = `
[
  {
    "isAndroid": false,
    "isMobile": false,
  },
  {
    "addressBarShadowPadding": 6,
    "devicePixelRatio": NaN,
    "deviceRectangles": {
      "bottomBar": {
        "height": 0,
        "width": 390,
        "x": 0,
        "y": 800,
      },
      "homeBar": {
        "height": 34,
        "width": 390,
        "x": 0,
        "y": 780,
      },
      "leftSidePadding": {
        "height": 0,
        "width": 0,
        "x": 0,
        "y": 0,
      },
      "rightSidePadding": {
        "height": 0,
        "width": 0,
        "x": 0,
        "y": 0,
      },
      "screenSize": {
        "height": 844,
        "width": 390,
      },
      "statusBar": {
        "height": 47,
        "width": 390,
        "x": 0,
        "y": 0,
      },
      "statusBarAndAddressBar": {
        "height": 47,
        "width": 390,
        "x": 0,
        "y": 0,
      },
      "viewport": {
        "height": 733,
        "width": 390,
        "x": 0,
        "y": 47,
      },
    },
    "fullPageScrollTimeout": 1500,
    "hideAfterFirstScroll": [],
    "innerHeight": NaN,
    "isAndroid": false,
    "isAndroidChromeDriverScreenshot": false,
    "isAndroidNativeWebScreenshot": false,
    "isIOS": false,
    "isLandscape": false,
    "screenHeight": NaN,
    "screenWidth": NaN,
    "toolBarShadowPadding": 6,
  },
]
`;

exports[`saveFullPageScreen > should handle NaN dimension values 3`] = `
[
  {
    "data": [
      "screenshot-1",
      "screenshot-2",
    ],
    "fullPageHeight": 2000,
    "fullPageWidth": 1200,
  },
  {
    "devicePixelRatio": NaN,
    "isLandscape": false,
  },
]
`;

exports[`saveFullPageScreen > should handle NaN dimension values 4`] = `
[
  {
    "isAndroid": false,
    "isMobile": false,
  },
  {
    "actualFolder": "/test/actual",
    "base64Image": "fullpage-screenshot-data",
    "disableBlinkingCursor": false,
    "disableCSSAnimation": false,
    "enableLayoutTesting": false,
    "fileName": {
      "browserName": "chrome",
      "browserVersion": "120.0.0",
      "deviceName": "desktop",
      "devicePixelRatio": NaN,
      "formatImageName": "{tag}-{logName}-{width}x{height}-dpr-{dpr}",
      "isMobile": false,
      "isTestInBrowser": true,
      "logName": "chrome",
      "name": "chrome",
      "outerHeight": NaN,
      "outerWidth": NaN,
      "platformName": "desktop",
      "platformVersion": "120.0.0",
      "screenHeight": NaN,
      "screenWidth": NaN,
      "tag": "test-fullpage",
    },
    "filePath": {
      "browserName": "chrome",
      "deviceName": "desktop",
      "isMobile": false,
      "savePerInstance": true,
    },
    "hideElements": [],
    "hideScrollBars": true,
    "isLandscape": false,
    "isNativeContext": false,
    "platformName": "desktop",
    "removeElements": [],
  },
]
`;

exports[`saveFullPageScreen > should handle undefined hideAfterFirstScroll, hideElements, and removeElements 1`] = `
[
  {
    "isAndroid": false,
    "isMobile": false,
  },
  {
    "addressBarShadowPadding": 6,
    "devicePixelRatio": 2,
    "deviceRectangles": {
      "bottomBar": {
        "height": 0,
        "width": 390,
        "x": 0,
        "y": 800,
      },
      "homeBar": {
        "height": 34,
        "width": 390,
        "x": 0,
        "y": 780,
      },
      "leftSidePadding": {
        "height": 0,
        "width": 0,
        "x": 0,
        "y": 0,
      },
      "rightSidePadding": {
        "height": 0,
        "width": 0,
        "x": 0,
        "y": 0,
      },
      "screenSize": {
        "height": 844,
        "width": 390,
      },
      "statusBar": {
        "height": 47,
        "width": 390,
        "x": 0,
        "y": 0,
      },
      "statusBarAndAddressBar": {
        "height": 47,
        "width": 390,
        "x": 0,
        "y": 0,
      },
      "viewport": {
        "height": 733,
        "width": 390,
        "x": 0,
        "y": 47,
      },
    },
    "fullPageScrollTimeout": 1500,
    "hideAfterFirstScroll": [],
    "innerHeight": 900,
    "isAndroid": false,
    "isAndroidChromeDriverScreenshot": false,
    "isAndroidNativeWebScreenshot": false,
    "isIOS": false,
    "isLandscape": false,
    "screenHeight": 1080,
    "screenWidth": 1920,
    "toolBarShadowPadding": 6,
  },
]
`;

exports[`saveFullPageScreen > should use Bidi screenshot when available and enableLegacyScreenshotMethod is false 1`] = `
{
  "devicePixelRatio": 2,
  "fileName": "test-fullpage.png",
}
`;

exports[`saveFullPageScreen > should use Bidi screenshot when available and enableLegacyScreenshotMethod is false 2`] = `
[
  {
    "browserInstance": {
      "isAndroid": false,
      "isMobile": false,
    },
    "origin": "document",
  },
]
`;

exports[`saveFullPageScreen > should use Bidi screenshot when available and enableLegacyScreenshotMethod is false 3`] = `
[
  {
    "isAndroid": false,
    "isMobile": false,
  },
  {
    "actualFolder": "/test/actual",
    "base64Image": "bidi-screenshot-data",
    "disableBlinkingCursor": false,
    "disableCSSAnimation": false,
    "enableLayoutTesting": false,
    "fileName": {
      "browserName": "chrome",
      "browserVersion": "120.0.0",
      "deviceName": "desktop",
      "devicePixelRatio": 2,
      "formatImageName": "{tag}-{logName}-{width}x{height}-dpr-{dpr}",
      "isMobile": false,
      "isTestInBrowser": true,
      "logName": "chrome",
      "name": "chrome",
      "outerHeight": 1000,
      "outerWidth": 1200,
      "platformName": "desktop",
      "platformVersion": "120.0.0",
      "screenHeight": 1080,
      "screenWidth": 1920,
      "tag": "test-fullpage",
    },
    "filePath": {
      "browserName": "chrome",
      "deviceName": "desktop",
      "isMobile": false,
      "savePerInstance": true,
    },
    "hideElements": [],
    "hideScrollBars": true,
    "isLandscape": false,
    "isNativeContext": false,
    "platformName": "desktop",
    "removeElements": [],
  },
]
`;

exports[`saveFullPageScreen > should use Bidi screenshot when available and userBasedFullPageScreenshot is false 1`] = `
{
  "devicePixelRatio": 2,
  "fileName": "test-fullpage.png",
}
`;

exports[`saveFullPageScreen > should use Bidi screenshot when available and userBasedFullPageScreenshot is false 2`] = `
[
  {
    "browserInstance": {
      "isAndroid": false,
      "isMobile": false,
    },
    "origin": "document",
  },
]
`;

exports[`saveFullPageScreen > should use Bidi screenshot when available and userBasedFullPageScreenshot is false 3`] = `
[
  {
    "isAndroid": false,
    "isMobile": false,
  },
  {
    "actualFolder": "/test/actual",
    "base64Image": "bidi-screenshot-data",
    "disableBlinkingCursor": false,
    "disableCSSAnimation": false,
    "enableLayoutTesting": false,
    "fileName": {
      "browserName": "chrome",
      "browserVersion": "120.0.0",
      "deviceName": "desktop",
      "devicePixelRatio": 2,
      "formatImageName": "{tag}-{logName}-{width}x{height}-dpr-{dpr}",
      "isMobile": false,
      "isTestInBrowser": true,
      "logName": "chrome",
      "name": "chrome",
      "outerHeight": 1000,
      "outerWidth": 1200,
      "platformName": "desktop",
      "platformVersion": "120.0.0",
      "screenHeight": 1080,
      "screenWidth": 1920,
      "tag": "test-fullpage",
    },
    "filePath": {
      "browserName": "chrome",
      "deviceName": "desktop",
      "isMobile": false,
      "savePerInstance": true,
    },
    "hideElements": [],
    "hideScrollBars": true,
    "isLandscape": false,
    "isNativeContext": false,
    "platformName": "desktop",
    "removeElements": [],
  },
]
`;

exports[`saveFullPageScreen > should use legacy method when Bidi is available but both userBasedFullPageScreenshot and enableLegacyScreenshotMethod are true 1`] = `
{
  "devicePixelRatio": 2,
  "fileName": "test-fullpage.png",
}
`;

exports[`saveFullPageScreen > should use legacy method when Bidi is available but both userBasedFullPageScreenshot and enableLegacyScreenshotMethod are true 2`] = `
[
  {
    "isAndroid": false,
    "isMobile": false,
  },
  {
    "addressBarShadowPadding": 6,
    "devicePixelRatio": 2,
    "deviceRectangles": {
      "bottomBar": {
        "height": 0,
        "width": 390,
        "x": 0,
        "y": 800,
      },
      "homeBar": {
        "height": 34,
        "width": 390,
        "x": 0,
        "y": 780,
      },
      "leftSidePadding": {
        "height": 0,
        "width": 0,
        "x": 0,
        "y": 0,
      },
      "rightSidePadding": {
        "height": 0,
        "width": 0,
        "x": 0,
        "y": 0,
      },
      "screenSize": {
        "height": 844,
        "width": 390,
      },
      "statusBar": {
        "height": 47,
        "width": 390,
        "x": 0,
        "y": 0,
      },
      "statusBarAndAddressBar": {
        "height": 47,
        "width": 390,
        "x": 0,
        "y": 0,
      },
      "viewport": {
        "height": 733,
        "width": 390,
        "x": 0,
        "y": 47,
      },
    },
    "fullPageScrollTimeout": 1500,
    "hideAfterFirstScroll": [],
    "innerHeight": 900,
    "isAndroid": false,
    "isAndroidChromeDriverScreenshot": false,
    "isAndroidNativeWebScreenshot": false,
    "isIOS": false,
    "isLandscape": false,
    "screenHeight": 1080,
    "screenWidth": 1920,
    "toolBarShadowPadding": 6,
  },
]
`;

exports[`saveFullPageScreen > should use legacy method when Bidi is available but both userBasedFullPageScreenshot and enableLegacyScreenshotMethod are true 3`] = `
[
  {
    "data": [
      "screenshot-1",
      "screenshot-2",
    ],
    "fullPageHeight": 2000,
    "fullPageWidth": 1200,
  },
  {
    "devicePixelRatio": 2,
    "isLandscape": false,
  },
]
`;

exports[`saveFullPageScreen > should use legacy method when Bidi is available but both userBasedFullPageScreenshot and enableLegacyScreenshotMethod are true 4`] = `
[
  {
    "isAndroid": false,
    "isMobile": false,
  },
  {
    "actualFolder": "/test/actual",
    "base64Image": "fullpage-screenshot-data",
    "disableBlinkingCursor": false,
    "disableCSSAnimation": false,
    "enableLayoutTesting": false,
    "fileName": {
      "browserName": "chrome",
      "browserVersion": "120.0.0",
      "deviceName": "desktop",
      "devicePixelRatio": 2,
      "formatImageName": "{tag}-{logName}-{width}x{height}-dpr-{dpr}",
      "isMobile": false,
      "isTestInBrowser": true,
      "logName": "chrome",
      "name": "chrome",
      "outerHeight": 1000,
      "outerWidth": 1200,
      "platformName": "desktop",
      "platformVersion": "120.0.0",
      "screenHeight": 1080,
      "screenWidth": 1920,
      "tag": "test-fullpage",
    },
    "filePath": {
      "browserName": "chrome",
      "deviceName": "desktop",
      "isMobile": false,
      "savePerInstance": true,
    },
    "hideElements": [],
    "hideScrollBars": true,
    "isLandscape": false,
    "isNativeContext": false,
    "platformName": "desktop",
    "removeElements": [],
  },
]
`;
